src/fifelse.c

```diff
@@ -373,29 +376,27 @@ SEXP fcaseR(SEXP na, SEXP rho, SEXP args) {
       }
     } break;
     case VECSXP: {
-      const SEXP *restrict pouts = SEXPPTR_RO(outs);
-      const SEXP pna = SEXPPTR_RO(na)[0];
+      // the default value of VECSXP is `NULL` so we don't need to explicitly
+      // assign the NA values as it does for other atomic types
+      const SEXP *restrict pthens;
+      if (!naout) pthens = SEXPPTR_RO(thens); // the content is not useful if out is NA_LOGICAL scalar
       for (int64_t j=0; j<len2; ++j) {
         const int64_t idx = imask ? j : p[j];
-        if (pcons[idx]==1) {
-          SET_VECTOR_ELT(ans, idx, pouts[idx & amask]);
+        if (pwhens[idx & conmask]==1) {
+          if (!naout) SET_VECTOR_ELT(ans, idx, pthens[idx & amask]);
         } else {
-          if (imask && nonna) {
-            SET_VECTOR_ELT(ans, idx, pna);
-          }
           p[l++] = idx;
         }
       }
     } break;
     default:
-      error(_("Type '%s' is not supported"), type2char(TYPEOF(outs)));
+      error(_("Type '%s' is not supported."), type2char(TYPEOF(ans)));
```

Comment:
I think we should mention what argument the "Type" is referring to.

Reply:
One thing here is getting a consistent argument to make translation easier:

https://github.com/search?q=repo%3ARdatatable%2Fdata.table+%2F%22Type+%27%25s%27+is+not+supported%2F&type=code

So I would move this to a follow-up where we try and improve that message more generally, maybe as "Type '%s' is not supported in argument %s".